hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroFileReaderFactory.java

```diff
@@ -45,11 +44,16 @@ protected HoodieFileReader newHFileFileReader(HoodieConfig hoodieConfig,
     if (isUseNativeHFileReaderEnabled(hoodieConfig)) {
       return new HoodieNativeAvroHFileReader(conf, path, schemaOption);
     }
-    CacheConfig cacheConfig = new CacheConfig(conf.unwrapAs(Configuration.class));
-    if (schemaOption.isPresent()) {
-      return new HoodieHBaseAvroHFileReader(conf, path, cacheConfig, HoodieStorageUtils.getStorage(path, conf), schemaOption);
+    try {
+      if (schemaOption.isPresent()) {
+        return (HoodieFileReader) ReflectionUtils.loadClass("org.apache.hudi.io.storage.HoodieHBaseAvroHFileReader",
+            new Class<?>[] {StorageConfiguration.class, StoragePath.class, Option.class}, conf, path, schemaOption);
+      }
+      return (HoodieFileReader) ReflectionUtils.loadClass("org.apache.hudi.io.storage.HoodieHBaseAvroHFileReader",
+          new Class<?>[] {StorageConfiguration.class, StoragePath.class}, conf, path);
+    } catch (HoodieException e) {
+      throw (IOException) e.getCause().getCause();
```

Comment:
Throw `new HoodieException("Cannot instantiate HoodieHBaseAvroHFileReader", e)`?

Reply:
fixed, but has to be ioexception because that is expected by one of the callers